### ISSUE TITLE
Route53 fallthrough to Route53 - Credential Bug

### ISSUE THREAD

**Issue by SuperFunks** [coredns/coredns#2205]
Hi there,

I'm trying to fallthrough route53 to another route53 section. Seems like that currently has a bug where the last profile in the credentials lines is used on all route53 sections. This of course gives access denied on the other zones and keeps CoreDNS from loading the zones.

Example (ID's have been changed):

```
.{
        route53 zone1.com.:Z2CDP34BQB3GIC {
                credentials 280846555434 /tmp/credentials
                fallthrough
        }
        route53 zone2.com.:Z3QSOTUJVXOESI {
                credentials 280846555434 /tmp/credentials
                fallthrough
        }
        route53 zone3.com.:Z2GB9FKQHZH7FN {
                credentials 188228134507 /tmp/credentials
        }
}
```

AccessDenied: User: arn:aws:sts::188228134507:assumed-role/ADFS-Viewer/coredns is not authorized to access this resource

Note that this works fine with the 2 first entries in there as they have the same credential profile - but adding the last one breaks the entire thing.

Hoping this could be fixed soon :)

Best regards,

SuperFunks

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[coredns/coredns] Route53 fallthrou..." ]
>Hi there,
>
>I'm trying to fallthrough route53 to another route53 section. Seems like that currently has a bug where the last profile in the credentials lines is used on all route53 sections. This of course gives access denied on the other zones and keeps CoreDNS from loading the zones.
>
>Example (ID's have been changed):
>
>.{
>        route53 zone1.com.:Z2CDP34BQB3GIC {
>                credentials 280846555434 /tmp/credentials
>                fallthrough
>        }
>        route53 zone2.com.:Z3QSOTUJVXOESI {
>                credentials 280846555434 /tmp/credentials
>                fallthrough
>        }
>        route53 zone3.com.:Z2GB9FKQHZH7FN {
>                credentials 188228134507 /tmp/credentials
>        }
>}
>
>AccessDenied: User: arn:aws:sts::188228134507:assumed-role/ADFS-Viewer/coredns is not authorized to access this resource
>
>Note that this works fine with the 2 first entries in there as they have the same credential profile - but adding the last one breaks the entire thing.
>
>Hoping this could be fixed soon :)

Should be fairly simple, but not all plugins allow multiple stanza in the 
corefile as this requires some administration the plugin's code, as in this case 
there the credentials get overwritten. Some plugins error on multiple use.

**Comment by yongtang** [comment]
I think appending credential providers will resolve the issue? Created #2218.

**Comment by canthefason** [comment]
I don't know if this would be a temporary solution for your problem, it might also be possible to configure it this way:

```
zone3.com. {
        route53 zone3.com.:Z2GB9FKQHZH7FN {
                credentials 188228134507 /tmp/credentials
        }
}

.{
        route53 zone1.com.:Z2CDP34BQB3GIC zone2.com.:Z3QSOTUJVXOESI {
                credentials 280846555434 /tmp/credentials
                fallthrough
        }
}
```

**Comment by sharmaansh21** [comment]
@yongtang @daniel-garcia  Hi if no one is working on this, can I look into it ?

**Comment by yongtang** [comment]
@sharmaanshul2102 That would be great! For some reasons I am not able to access multiple AWS account to validate and debug the issue (so didn't make any progress before). If you could provide the fix and PR that would be wonderful. Let me know if there is any help I can provide.

**Comment by sharmaansh21** [comment]
I worked on the fix, and I resolved the problem need to test more thoroughly. Will send PR next week.

**Comment by sharmaansh21** [comment]
@yongtang https://github.com/coredns/coredns/pull/2859

**Comment by miekg** [comment]
think this should now work?

**Comment by sharmaansh21** [comment]
@miekg yes :)

**Comment by yongtang** [comment]
@SuperFunks Looks like the issue has been fixed in #2859. Will close this issue but feel free to re-open if issue persist.
